File: sem4/mpp/lab11-12/frontend/src/app/reporting/service/reporting.service.ts
import { Injectable } from '@angular/core';
import {Observable} from 'rxjs';
import {HttpClient} from '@angular/common/http';

@Injectable({
  providedIn: 'root'
})
export class ReportingService {

  SERVER_ROOT_URL = 'http://localhost:8080';

  constructor(private http: HttpClient) { }

  getTopClients(): Observable<object[]>
  {
    return this.http.get<object[]>(`${this.SERVER_ROOT_URL}/top-clients?limit=3`);
  }

  getBadRenters(): Observable<object[]>
  {
    return this.http.get<object[]>(`${this.SERVER_ROOT_URL}/bad-renters`);
  }

  getRentalsByGenre(): Observable<object[]>
  {
    return this.http.get<object[]>(`${this.SERVER_ROOT_URL}/rentals-genre`);
  }
}

File: sem4/mpp/lab11-12/backend/src/main/java/server/services/reporting/ReportingService.java
package server.services.reporting;

import server.entities.Client;
import server.entities.GenreEnum;
import javafx.util.Pair;

import java.util.List;

public interface ReportingService
{
    /**
     * Fetch top `limit` clients by number of rentals.
     * @return List of pairs containing client and number of rentals.
     */
    List<Pair<Client, Integer>> getTopClients(int limit);

    /**
     * Get clients who had at least one rental for more than 30 days.
     * @return List of bad renters.
     */
    List<Client> getBadRenters();

    /**
     * See how many rentals were made for each genre.
     * @return List of pairs consisting of genre and count.
     */
    List<Pair<GenreEnum, Integer>> getRentalsByGenre();
}
